### PR TITLE
Session upload support

### PR DESCRIPTION
**Pull request by saileru**
-f/--format is now optional iff the format is instead specified in the input file via _[format]_. If any files are given as plain arguments, the format must, however, be specified.

Example:
```
folder/coverage_upload.simple
[jacoco]
pattern1/**.xml
pattern2/**.xml
[findbugs]
pattern1/**.findbugs.xml
pattern2/**.findbugs.xml
```
In this case, '-f simple' must be specified as default format. The first entry in the input file does not have a format and, thus, needs to fall back to the default format.

Currently, there is one separate request for every upload format as I am not sure if/how it is possible to combine different upload formats into one request.

I did some basic - however not exhaustive - testing. So, there might be bugs left in certain cases.

### CODE REVIEW

**Review by karottenreibe — CHANGES_REQUESTED** [review]
great work! some comments :-)

File: src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java [hunk 1]
```diff
@@ -71,9 +81,14 @@ public void validate(ArgumentParser parser) throws ArgumentParserException {
...
-                throw new ArgumentParserException("You must either specify the paths of the coverage files as plain " +
-                        "arguments or provide them in an input file, see help for more information", parser);
+            if (files.isEmpty() && input == null) {
+                throw new ArgumentParserException("No report files provided." +
+                        " You must either specify the paths of the coverage files as command line" +
+                        " arguments or provide them in an input file via --input", parser);
+            }
+
+            if (format == null) {
+                throw new ArgumentParserException("Please provide the default report format with --format", parser);
```
Comment (saileru) [inline src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java:91]: I made the format argument mandatory. It may not be necessary when all patterns have a corresponding [format] in the input file. However, it is much simpler if we do not have to check later if the default format is set.

Another option would be to automatically set a fallback format if the user does not specify one. This might, however, cause unexpected behavior when the user forgets to specify the format and files are then uploaded with the wrong fallback format. 

What do you think?

Comment (karottenreibe) [reply src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java:91]: That works for me, but instead of checking yourself manually please use the argument parser library to do the required check like here: https://github.com/cqse/teamscale-upload/blob/e97ab41ed16c16c9a47c3e47b3155a80bfc828c8/src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java#L125
`.required(true)`

This way it's consistent with the other parameters and highlighted as required in `--help`

Comment (karottenreibe) [reply src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java:91]: Actually, while reading the code I had another idea: how about we make it mandatory to specify a format in the file and only apply the `--format` to stuff given on the command line?
That would make the code simpler I think and we can clearly document this in `--help`. What do you think?

File: src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java [hunk 2]
```diff
@@ -185,24 +196,72 @@ private static Input parseArguments(String[] args) {
...
+                .addPathSegments("api/projects").addPathSegment(input.project)
+                .addPathSegments("external-analysis/session")
+                .addQueryParameter("partition", input.partition);
+
+        HttpUrl url = builder.build();
 
-        for (String format : formatToFileNames.keySet()) {
-            sendRequestForFormat(input, format, formatToFileNames.get(format));
+        // Empty body
+        RequestBody body = RequestBody.create(null, new byte[0]);
```
Comment (karottenreibe) [inline src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java:224]: extract as constant? it also appears below again

File: src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java [hunk 3]
```diff
@@ -185,24 +196,72 @@ private static Input parseArguments(String[] args) {
...
+                .url(url)
+                .post(body)
+                .build();
+
+        System.out.print("Opening upload session...");
+        String sessionId = sendRequest(input, url, request);
+        if (sessionId == null) {
+            fail("Could not open session.");
         }
+        System.out.println("Session id: " + sessionId);
```
Comment (karottenreibe) [inline src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java:237]:
```suggestion
        System.out.println("Session ID: " + sessionId);
```

File: src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java [hunk 4]
```diff
@@ -185,24 +196,72 @@ private static Input parseArguments(String[] args) {
...
+
+        // Empty body
+        RequestBody body = RequestBody.create(null, new byte[0]);
+
+        Request request = new Request.Builder()
+                .header("Authorization", Credentials.basic(input.username, input.accessKey))
+                .url(url)
+                .post(body)
+                .build();
+        System.out.print("Closing upload session...");
```
Comment (karottenreibe) [inline src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java:257]:
```suggestion
        System.out.println("Closing upload session");
```

File: src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java [hunk 5]
```diff
@@ -243,17 +305,30 @@ private static void sendRequestForFormat(Input input, String format, List<String
                 .post(requestBody)
                 .build();
 
+        System.out.print("Sending upload for format " + format + "...");
```
Comment (karottenreibe) [inline src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java:308]:
```suggestion
        System.out.println("Uploading reports for format " + format);
```

File: src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java [hunk 6]
```diff
@@ -243,17 +305,30 @@ private static void sendRequestForFormat(Input input, String format, List<String
...
+        } catch (UnknownHostException e) {
+            fail("The host " + url + " could not be resolved. Please ensure you have no typo and that" +
+                    " this host is reachable from this server. " + e.getMessage());
+        } catch (ConnectException e) {
+            fail("The URL " + url + " refused a connection. Please ensure that you have no typo and that" +
+                    " this endpoint is reachable and not blocked by firewalls. " + e.getMessage());
         } finally {
             // we must shut down OkHttp as otherwise it will leave threads running and
             // prevent JVM shutdown
             client.dispatcher().executorService().shutdownNow();
```
Comment (karottenreibe) [inline src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java:328]: this should only happen after the last request, so we can't do this in `sendRequest`, which is called multiple times. Please move this top-level into the `main` method

File: src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java [hunk 7]
```diff
@@ -243,17 +305,30 @@ private static void sendRequestForFormat(Input input, String format, List<String
...
+    }
+
+    private static String sendRequest(Input input, HttpUrl url, Request request) throws IOException {
         OkHttpClient client = OkHttpClientUtils.createClient(input.validateSsl);
 
         try (Response response = client.newCall(request).execute()) {
             handleCommonErrors(response, input);
-            System.out.println("Upload to Teamscale successful for format " + format);
+            System.out.println("Successful");
+            return response.body().string();
```
Comment (karottenreibe) [inline src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java:318]: if I remember correctly, `body()` may already be `null` which we should handle

File: src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java [hunk 8]
```diff
@@ -243,17 +305,30 @@ private static void sendRequestForFormat(Input input, String format, List<String
...
 
+        System.out.print("Sending upload for format " + format + "...");
+        sendRequest(input, url, request);
+    }
+
+    private static String sendRequest(Input input, HttpUrl url, Request request) throws IOException {
         OkHttpClient client = OkHttpClientUtils.createClient(input.validateSsl);
 
         try (Response response = client.newCall(request).execute()) {
             handleCommonErrors(response, input);
```
Comment (karottenreibe) [inline src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java:316]: If I remember correctly, handleCommonErrors actually handles _all_ errors. Should we rename it to `handleErrors`?

File: src/test/java/com/cqse/teamscaleupload/NativeImageIT.java [hunk 9]
```diff
@@ -88,13 +93,21 @@ public void insufficientPermissions() {
...
-    public void successfulUpload() {
+    public void successfulSingleFormatUpload() {
         ProcessUtils.ProcessResult result = runUploader(new Arguments());
         assertThat(result.exitCode)
                 .describedAs("Stderr and stdout: " + result.stdoutAndStdErr)
                 .isZero();
     }
 
+    @Test
+    public void successfulSessionUpload() {
```
Comment (karottenreibe) [inline src/test/java/com/cqse/teamscaleupload/NativeImageIT.java:104]: if I understand correctly, the single format case also uses a session, so maybe rename to `successfulMultiFormatUpload`?

**Review by karottenreibe — CHANGES_REQUESTED** [review]
some comments

File: src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java [hunk 10]
```diff
@@ -185,24 +196,72 @@ private static Input parseArguments(String[] args) {
...
+            sendRequestForFormat(input, format, formatToFileNamesWrapper.getFilesForFormat(format), sessionId);
         }
+        closeSession(input, sessionId);
+    }
+
+    private static String openSession(Input input) throws IOException {
+        HttpUrl.Builder builder = input.url.newBuilder()
+                .addPathSegments("api/projects").addPathSegment(input.project)
+                .addPathSegments("external-analysis/session")
+                .addQueryParameter("partition", input.partition);
```
[the remark this answers is not in the document]
Comment (saileru) [reply src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java:237]: Sure, makes sense. I moved this to the call for the session start.

File: coverage_files/input_file [hunk 11]
```diff
@@ -1,4 +1,7 @@
-coverage_files\test*.simple
+
+
 [jacoco]
+coverage_files/test*.simple
+
 [simple]
-coverage_files\coverage.simple
\ No newline at end of file
+coverage_files/coverage.simple
```
Comment (karottenreibe) [inline coverage_files/input_file:7]: why did you change this to a forward slash? can we safely handle windows paths? Windows users probably expect backslashes to work

File: src/main/java/com/cqse/teamscaleupload/ReportPatternManager.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+        String currentFormat = null;
+        for (String line : nonEmptyLines) {
+            Matcher formatPatternMatcher = FORMAT_PATTERN.matcher(line);
+            if (formatPatternMatcher.matches()) {
+                currentFormat = formatPatternMatcher.group(1).toUpperCase();
+                formatToFileNames.computeIfAbsent(currentFormat, k -> new HashSet<>());
+            } else {
+                formatToFileNames.get(currentFormat).add(line);
+            }
+        }
```
Comment (karottenreibe) [inline src/main/java/com/cqse/teamscaleupload/ReportPatternManager.java:55]: we should also `fail()` here if after parsing, we have 0 actual patterns. In that case there's nothing to upload and this is probably a mistake by the user

File: src/main/java/com/cqse/teamscaleupload/ReportPatternManager.java [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    /**
+     * Return a set of all file patterns for the given format.
+     * Returns null if the given format does not exist.
+     */
+    public Set<String> getPatternsForFormat(String format) {
```
Comment (karottenreibe) [inline src/main/java/com/cqse/teamscaleupload/ReportPatternManager.java:62]: the caller doesn't currently check for `null`. I think it would also be simpler if we returned an empty set here instead of `null` by using `getOrDefault()`. What do you think?

File: src/main/java/com/cqse/teamscaleupload/ReportPatternManager.java [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+
+    /**
+     * Returns all formats which are used in this session.
+     */
+    public Set<String> getAllUsedFormats() {
+        return formatToFileNames.keySet();
+    }
+
+    /**
+     * Adds the file patters for the given format.
```
Comment (karottenreibe) [inline src/main/java/com/cqse/teamscaleupload/ReportPatternManager.java:74]:
```suggestion
     * Adds the file patterns for the given format.
```

File: src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java [hunk 15]
```diff
@@ -26,9 +26,12 @@
 import java.nio.file.Paths;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Set;
 
 public class TeamscaleUpload {
 
+    public static final RequestBody EMPTY_BODY = RequestBody.create(null, new byte[0]);
```
Comment (karottenreibe) [inline src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java:33]: can you please resolve the TS findings?

Comment (karottenreibe) [reply src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java:33]: We can probably make some stuff private to resolve the interface comment warnings

File: src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java [hunk 16]
```diff
@@ -168,11 +171,11 @@ private static Input parseArguments(String[] args) {
         parser.epilog("INPUTFILE" +
                 "\n" +
                 "\n" +
-                "The file you provide via --input consists of file patterns in the same format as used on the command line" +
-                " and optionally sections that specify additional report file formats." +
-                " The entries in the file are separated by line breaks. Uses the format specified with --format," +
-                " unless you overwrite the format explicitly for a set of patterns:\n\n" +
-                "pattern1/**.simple\n" +
+                "The input file allows to upload report files for different file patterns in one upload session." +
```
Comment (karottenreibe) [inline src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java:174]:
```suggestion
                "The input file allows to upload multiple report files for different formats in one upload session." +
```

File: src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java [hunk 17]
```diff
@@ -168,11 +171,11 @@ private static Input parseArguments(String[] args) {
...
+                " Each section of reports must start with a specification of the report format. The report file" +
+                " patterns have the same format as used on the command line." +
+                " The entries in the file are separated by line breaks.\n\n" +
+                "Example:\n\n" +
```
Comment (karottenreibe) [inline src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java:178]: can you add another section with a different format to the example please? just to make it explicit that you can use the format header multiple times.

File: src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java [hunk 18]
```diff
@@ -168,11 +171,11 @@ private static Input parseArguments(String[] args) {
...
                 "\n" +
-                "The file you provide via --input consists of file patterns in the same format as used on the command line" +
-                " and optionally sections that specify additional report file formats." +
-                " The entries in the file are separated by line breaks. Uses the format specified with --format," +
-                " unless you overwrite the format explicitly for a set of patterns:\n\n" +
-                "pattern1/**.simple\n" +
+                "The input file allows to upload report files for different file patterns in one upload session." +
+                " Each section of reports must start with a specification of the report format. The report file" +
+                " patterns have the same format as used on the command line." +
+                " The entries in the file are separated by line breaks.\n\n" +
```
Comment (karottenreibe) [inline src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java:177]:
```suggestion
                " The entries in the file are separated by line breaks. Blank lines are ignored.\n\n" +
```

File: src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java [hunk 19]
```diff
@@ -196,71 +199,95 @@ private static Input parseArguments(String[] args) {
...
 
-    private static void performUpload(FormatToFileNamesWrapper formatToFileNamesWrapper, Input input) throws IOException, AgentOptionParseException {
-        String sessionId = openSession(input);
-        for (String format : formatToFileNamesWrapper.getAllAvailableFormats()) {
-            sendRequestForFormat(input, format, formatToFileNamesWrapper.getFilesForFormat(format), sessionId);
+    private static void performUpload(OkHttpClient client, ReportPatternManager reportPatternManager, Input input) throws IOException, AgentOptionParseException {
+        String sessionId = openSession(client, input);
+        for (String format : reportPatternManager.getAllUsedFormats()) {
+            Set<String> patternsForFormat = reportPatternManager.getPatternsForFormat(format);
+            if (patternsForFormat.isEmpty()) {
```
Comment (karottenreibe) [inline src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java:224]: can we move this validation step into the `PatternManager`? Otherwise there's a possibility that we upload 9/10 formats and then the 10th fails and the session is never closed. In that case I'd prefer to fail early during input file parsing instead.

File: src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java [hunk 20]
```diff
@@ -196,71 +199,95 @@ private static Input parseArguments(String[] args) {
...
+        } else if (input.timestamp != null) {
+            builder.addQueryParameter("t", input.timestamp);
+        } else if (input.autoDetectedCommit) {
+            String commit = detectCommit();
+            if (commit == null) {
+                fail("Failed to automatically detect the commit. Please specify it manually via --commit or --timestamp");
+            }
+            builder.addQueryParameter("revision", commit);
+        } else {
+            builder.addQueryParameter("t", "master:HEAD");
```
Comment (karottenreibe) [inline src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java]: this assumes Git and a default branch name of `master`. Can we somehow make this work with any default branch name, e.g. `trunk` from SVN?

**Review by karottenreibe — CHANGES_REQUESTED** [review]

File: coverage_files/input_file (hunk 11 above)
Comment (saileru) [reply coverage_files/input_file:7]: This failed on the build server for Linux: https://travis-ci.com/github/cqse/teamscale-upload/jobs/396920590

As Windows seems to support both slashes as path separators, I would suggest we normalize all slashes to forward slashes. This way, it should work for both.

Another solution would be

File: src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java (hunk 17 above)
Comment (saileru) [reply src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java:178]: I do not quite understand what you mean. The example already contains multiple formats (`[jacoco]` and `[findbugs]`).

Or do you mean something else?

File: src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java (hunk 20 above)
Comment (saileru) [reply src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java]: I am not really sure, how we would do that. Do you have any suggestions?

If this requires a non-trivial change, I would prefer to move this to a separate issue, so that we can get the session support merged.

Comment (karottenreibe) [reply src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java]: I had a quick look at the TS code and I think just passing `"HEAD"` should choose the default branch. Can you quickly check if that works? If not, please create a separate issue :-) :+1:

File: src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java (hunk 17 above)
Comment (karottenreibe) [reply src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java:185]: Aaah sorry, the GitHub Diff had collapsed the rest of the example. My bad, sorry! :-)

File: src/main/java/com/cqse/teamscaleupload/ReportPatternManager.java [hunk 21]
```diff
@@ -13,23 +14,49 @@
 import java.util.stream.Collectors;
 
 /**
- * This class manages the given file formats and
- * corresponding report file patterns.
+ * This class provides the functionality to resolve the report file patterns and formats
+ * from the command line to a map from report formats to actual files which can then be
+ * used for the upload to Teamscale.
```
Comment (karottenreibe) [inline src/main/java/com/cqse/teamscaleupload/ReportPatternManager.java:19]: You changed this class so that it now only contains static methods. I'd thus prefer we rename this `ReportPatternUtils` as it no longer "manages" anything ;-)

Comment (karottenreibe) [reply src/main/java/com/cqse/teamscaleupload/ReportPatternManager.java:19]: Alternative: make the methods non-static and re-introduce the field you had previously for the List-Map. Then you don't have to pass that to every private utility method. It's up to you.

File: src/main/java/com/cqse/teamscaleupload/ReportPatternManager.java [hunk 22]
```diff
@@ -13,23 +14,49 @@
...
+    public static Map<String, Set<File>> resolveInputFilePatters(Path inputFile, List<String> filePatterns, String format)
+            throws IOException, AgentOptionParseException {
+        Map<String, Set<String>> formatToFilePatterns = new HashMap<>();
+
+        if (inputFile != null) {
+            formatToFilePatterns = addReportFilePatternsFromInputFile(inputFile);
+        }
+        addFilePatternsForFormatOnCommandLine(formatToFilePatterns, filePatterns, format);
+
+        // Resolve all file patterns to the actual files for all given formats.
```
Comment (karottenreibe) [inline src/main/java/com/cqse/teamscaleupload/ReportPatternManager.java:44]: extract this part to its own method? inline comments for me are hints that this is something that should be encapsulated in a funciton with a speaking name instead

File: src/main/java/com/cqse/teamscaleupload/ReportPatternManager.java [hunk 23]
```diff
@@ -13,23 +14,49 @@
...
+        }
+
+        return formatToFiles;
+    }
+
+    /**
+     * Returns a map from report format to file patterns which are read from the given input file.
+     * In the end, it is verified that no format has an empty set of patterns ({@link #validatePatternsForFormats(Map)})
+     */
+    private static Map<String, Set<String>> addReportFilePatternsFromInputFile(Path input) throws IOException {
```
Comment (karottenreibe) [inline src/main/java/com/cqse/teamscaleupload/ReportPatternManager.java:58]: `add` seems weird since this doesn't add anything anymore. Instead it returns a map. Can you rename the function please? E.g. `parseInputFile` or something along those lines.

**Review by karottenreibe — CHANGES_REQUESTED** [review]
some minor comments

File: src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java (hunk 20 above)
Comment (saileru) [reply src/main/java/com/cqse/teamscaleupload/TeamscaleUpload.java]: I adjusted the value to "HEAD" and the upload for our git project still works. 👍 
We might at some point need to test this for SVN and others. But, for now, this should do.

**Comment by saileru** [comment]
Thanks, I applied your suggestions 🙂

**Review by karottenreibe — APPROVED** [review]

**Comment by karottenreibe** [comment]
Good work and sorry for the long round-trip times with my reviews!
